hugr-passes/src/dataflow/partial_value.rs

```diff
@@ -0,0 +1,685 @@
+use ascent::lattice::BoundedLattice;
+use ascent::Lattice;
+use hugr_core::ops::Value;
+use hugr_core::types::{ConstTypeError, SumType, Type, TypeEnum, TypeRow};
+use itertools::{zip_eq, Itertools};
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+
+/// Trait for an underlying domain of abstract values which can form the *elements* of a
+/// [PartialValue] and thus be used in dataflow analysis.
+pub trait AbstractValue: Clone + std::fmt::Debug + PartialEq + Eq + Hash {
+    /// Computes the join of two values (i.e. towards `Top``), if this is representable
+    /// within the underlying domain.
+    /// Otherwise return `None` (i.e. an instruction to use [PartialValue::Top]).
+    ///
+    /// The default checks equality between `self` and `other` and returns `self` if
+    /// the two are identical, otherwise `None`.
+    fn try_join(self, other: Self) -> Option<Self> {
+        (self == other).then_some(self)
+    }
+
+    /// Computes the meet of two values (i.e. towards `Bottom`), if this is representable
+    /// within the underlying domain.
+    /// Otherwise return `None` (i.e. an instruction to use [PartialValue::Bottom]).
+    ///
+    /// The default checks equality between `self` and `other` and returns `self` if
+    /// the two are identical, otherwise `None`.
+    fn try_meet(self, other: Self) -> Option<Self> {
+        (self == other).then_some(self)
+    }
+}
+
+/// Represents a sum with a single/known tag, abstracted over the representation of the elements.
+/// (Identical to [Sum](hugr_core::ops::constant::Sum) except for the type abstraction.)
```

Comment:
Commit 7772dfc04b6b99580f9b11a962c98cd3f174ef04 combines these two/dedups, but at the cost of having to manually implement serde Serialize/Deserialize, so see what you think